### PR TITLE
Add standard upload method variables

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 
This PR fixes some deficiencies in the upload method system which are starting to become more of a problem in the newest version of Mbed.

One issue has been around for a while: if you want to set the serial number of the device to program, it requires a different way to do that for each upload method: PYOCD_PROBE_UID, STLINK_PROBE_SN, etc.  All of these are set to the same "type" of thing (the USB serial number of the probe to use), and yet the name is different for each upload method you use. This is silly, and these variables should be unified into one common parameter to set.  This change does that.

The other issue is more recent.  @ccli8 's recent PRs have switched many tools from using the elf file to the bin/hex file to upload, and this is good because it's required for TFM targets and targets with post build hooks that modify the image to work.  However, when uploading a bin file, tools need to be told what address to load it at.  This information hasn't existed in Mbed until recently, and it was not being provided to the upload methods at all.

This actually caused a bug I noticed in testing, where flashing code using OpenOCD does not work (just silently left the old version of the code on there) because it defaults to loading to address 0x0, and that is not where some targets put their flash image.

To fix these issue, I've added two new "common" upload method options: `MBED_UPLOAD_SERIAL_NUMBER` and `MBED_UPLOAD_BASE_ADDR`.  The former is intended to be set by the user, and defaults to empty though it will pick up the value from the legacy options it replaces if any of those are set.  The latter is initialized based on the configured first flash bank start address (MBED_CONFIGURED_ROM_START) but can also be set to a custom address in the user's upload method configuration.  I also made `MBED_GDB_PORT` into a proper option if you want to change the port that the GDB server runs on.

#### Impact of changes 
- Upload method UID/serial number arguments condensed into one common one, `MBED_UPLOAD_SERIAL_NUMBER`.  However, legacy variables are still supported and will print a warning if used.
- New option to specify where a bin file is loaded to in the target's memory, `MBED_UPLOAD_BASE_ADDR`
- Fixed bug where some upload methods would always flash the bin/hex file to address 0, even for devices (e.g. STM32 micros) where that is not a valid flash location
- Improved error message if you have an old build directory from before the memory bank information change landed that needs to be reconfigured

### Documentation 

Will update the upload method documentation once this is ready to be merged

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [X] Tests / results supplied as part of this PR
    
So far I have verified this with st-flash, OpenOCD, pyocd, and LinkServer, and code can be flashed and runs OK.
----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: targets/upload_method_cfg/K64F.cmake [hunk 1]
```diff
@@ -3,11 +3,10 @@
 # include app.cmake and where you add mbed os as a subdirectory.
 #
 # Notes:
-# 1. PyOCD did not actually work in my testing as of Apr 2024, though this device is supposed to be supported
-# 2. Be sure to update the DAPLink firmware on the board via these instructions: https://os.mbed.com/blog/entry/DAPLink-bootloader-update/
-# 3. OpenOCD 0.12 flashes this device perfectly and can enter a debug session, but cannot hit breakpoints
-# 4. LinkServer can both flash and debug, so it's the recommended upload method for this device.
-# 5. LinkServer does appear to have a bug where it doesn't map the peripheral registers as valid memory, so you can't
+# 1. Be sure to update the DAPLink firmware on the board via these instructions: https://os.mbed.com/blog/entry/DAPLink-bootloader-update/
```
Comment (multiplemonomials) [inline targets/upload_method_cfg/K64F.cmake:6]: I retested this and pyocd works fine now 🤷

**Comment by multiplemonomials** [comment]
@JohnK1987 @ccli8 could one of you review and approve? Thx!

**Review by ccli8 — APPROVED** [review]
Except confused on `MBED_UPLOAD_BASE_ADDR` used with hex file, no other concern.

**Review by JohnK1987 — APPROVED** [review]

**Comment by multiplemonomials** [comment]
Updated the wiki page here with docs: https://github.com/mbed-ce/mbed-os/wiki/Upload-Methods.  Made sure to add a warning about the base address option when hex files are used.

**Pull request merged**